README.md
# Serial Test Program
This program using win32API to Tx/Rx from serial COM port under windows OS

## Overview
---
This document is for developers to get up and running.

## Getting Started
---
### Requirements
1. Windows operating system
2. GNU C compiler (e.g. gcc)
3. USB to TTL dongle
4. Your testing device

### Set up
- Create your own main.c or using the given example.c 
- Include serial.h and serial.c (if needed)
- Define your own testing commands base on the Interface Control Document (ICD) of the target device

### In main()
Make sure your code in main() follows the sequence below.

1. `initComPort()` 
   
   Passing number of COM port and baudrate as function arguments.
   
2. `GC(type)()`
 
   1. generate command, usually header
   2. generate command, usually the main content of testing command
   3. generate command, usually footer
   > Note: This step depends on user's testing command, if the user need advance seting of command (e.g lsb or msb) see `command_t` in serial.h
   
   > Note: While using `GC(ascii)`, don't forget add one more byte at the argument `cpn_len` (due to #define "abc" is a string in C, there is a '\0' at the end of string)


3. `printCommand()` 
    > Note: It is not necessary to print out the command, it's for user to check and debug.

4. `transmit()`
   
    There is no argument needed while calling this function. Number of bytes been sent will print on terminal.

5. `receive()`

    There is no argument needed while calling this function. The terminate condition of this function is when it received a 0x0D byte ('CR' in ascii code) or timeout. Number of bytes received and received message will print on terminal.
    

6. `GC(clear)(_command_)`

    Always remember to free the memory space after using it.

7. `closePort()`

    Don't forget close the port you just opened.

### Compile and Run
Open the terminal at the same directory with your main.c
```
gcc -o main main.c
./main.exe
```
Then the user should see the command has been sent and the message received from testing device.\


## History Version
---
see `change_log.md`

  
## License
---
There is no license or copyright, just use it :)

CHANGELOG.md
# Serial

Using C language(WinAPI) to access serial port of PC

---------------------------------------------------------------
## Unreleased v1.1.2 (####.##.##)
### New
### Changed 
### Discarded
### Removed
### Fixed
### Known Issues
### Notes
### Reference

---------------------------------------------------------------
## Serial v1.2.0 (2023.01.11)
### Removed
- remove `_command_` argument in all GC()

---------------------------------------------------------------
## Serial v1.1.2 (2023.01.11)
### New
- selectable print receive format (hex/ascii)
- receive(1) = print ascii, else = print hex

---------------------------------------------------------------
## Serial v1.1.1 (2023.01.03)
### Changed 
- change print hex in `receive()`
### Removed
- remove `cpn_len`-1 in GC(ascii), now how many words is how many bytes input

---------------------------------------------------------------
## Serial v1.1 (2022.10.19)
### New
- update README.md
### Changed
- change `ReadIntervalTimeout` to 50 ms, and ReadTotalTimeoutConstant to 500 ms
### Discarded
- The 0x0D terminate in `ReadFile()`

---------------------------------------------------------------
## Serial v1.0 (2022.10.17)
### New
- add README.md
### Known Issues
- Need add more description in GC(ascii)
- Some device may have 0x0D in reply messages, can't use 0x0D as terminate condition while reading messages

---------------------------------------------------------------
## Serial v1.0 (2022.10.13)
### Notes
- v1.0 == v0.14

---------------------------------------------------------------
## Serial v0.14 (2022.10.13)
### New
- set rx timeout

---------------------------------------------------------------
## Serial v0.13 (2022.10.13)
### New
- add gen_command_crc() for CRC32
- add comment to all functions
### Changed 
- change gen_command_free() to gen_command_clear()
### Known Issues
- Some USB dongle may get error 0x57 at SetCommMask


---------------------------------------------------------------
## Serial v0.12 (2022.10.13)
### New
- can let user customize command
### Changed 
- merge gen_cmd.h into serial.h
- merge gen_cmd.c into serial.c

---------------------------------------------------------------
## Serial v0.11 (2022.10.12)
### New
- put number of port as param into initComPort() 
- add stderr fprintf to each return status
### Fixed
- Fixed all bugs in v0.10
### Notes
- DO NOT redefine a global variable in functions.
### Reference
- [SetCommState function (winbase.h)](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-setcommstate)
- [System Error Codes](https://learn.microsoft.com/zh-tw/windows/win32/debug/system-error-codes--0-499-)
- [CreateFileA function (fileapi.h)](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilea)

---------------------------------------------------------------
## Serial v0.10 (2022.10.09 - 2022.10.10)
### New
- modulize the serial_v0.9.c to serial.h and serial.c
### Changed 
- change command_body to body in command_t
- change command_len to length in command_t
### Note
- use extern to define global variable in .h file
- don't use send() as function name, there is another function called send() in "windows.h"

---------------------------------------------------------------
## Serial v0.9 (2022.10.09)
### Changed 
- move serial param and `rx_buffer_size` to #define
### Removed
- x.x.Txx is for TRX-U
- x.x.Axx is for APRS
- x.x.Xxx is for X-band
- before 0.7 is for X-band
- e.g.  
    - 0.8.X01 is version 01 for Xband using v 0.8 serial test
    - 0.8.T03 is version 03 for TRX-U using v 0.8 serial test

---------------------------------------------------------------
## Serial v0.8.T03 (2022.06.02)
### Known Issue
- ascii string will lead to error code in `gen_command`

---------------------------------------------------------------
## Serial v0.8.T02 (2022.06.02)
### Fixed
- Program will freeze when `rx_buffer` is larger then received
### Notes
X.X.2 is for TRX-U

---------------------------------------------------------------
## Serial v0.8.T01 (2022.06.01)
### New
- header and command
### Discarded
- TRX-U doesn't need CRC
### Known Issues
- Program will freeze when `rx_buffer` is larger then received number

---------------------------------------------------------------
## Serial v0.8 (2022.04.20)
### New
### Changed 
- count command_len in gen_command()
### Deprecated
### Removed
### Fixed
### Known Issues
### Notes
### Reference

---------------------------------------------------------------
## Serial v0.7.X02 (2022.04.20)
### Changed
- declare command as plain command
### Notes
- Also no reply from X-Band.

---------------------------------------------------------------
## Serial v0.7.X01 (2022.04.20)
### Changed
- change to MSB
### Notes
- Also no reply from X-Band.

---------------------------------------------------------------
## Serial v0.7.A01 (2022.03.24)
### New
### Changed
- Calculate CRC first then change to LSB
### Deprecated
### Removed
### Fixed
### Known Issues
### Notes
- Accroding to EnduroSat document(email) this is not correct way
### Reference

---------------------------------------------------------------

## Serial v0.7 (2022.03.23)
### New
- SetCommState()
- Some print for error handling (GetLastError())
### Changed
### Deprecated
### Removed
### Fixed
- Now can successfully open COM port
- Now can successfully set comm state
- Now can successfully change baudrate (verified by logic analyzer)
### Known Issues
### Notes
- Accroding to EnduroSat document, the command is correct
### Reference
[CreatFile() Error:5](https://blog.csdn.net/qq_37376568/article/details/96741297)

---------------------------------------------------------------
## Serial v0.6 (2022.03.14)
### New
- setting Rx and receiving data

### Changed
### Deprecated
### Removed
### Fixed
### Known Issues
- By logic analyzer, found cannot change baudrate
- Opening COM port failed (hComm failed)
- Doesn't have SetCommState()
- Set comm state failed
### Notes

main.c
#include "serial.h"
#include "serial.c"

/* ================== Example commands ================== */

#define HEADER  "NCU+"
#define STATUS  "STATUS=?"
#define BEACON  0x9487
#define FOOTER  0x0D0A       // "CRLF" in ascii


/* ================== Example codes ================== */
int main (void){

    /* ------------------ Initialize Serial COM Port ------------------ */
    initComPort(3, 9600);    // (port number, baudrate)
    

    /* ------------------ Set Commands ------------------ */
    GC(init)();
    GC(ascii)(HEADER, 4);
    GC(ascii)(STATUS, 8);
    // GC(hex)(BEACON, 2);
    // GC(crc)();
    GC(hex)(FOOTER, 2);

    /* ------------------ Print Commands ------------------ */
    printCommand();
    
    /* ------------------ Transmit Commands ------------------ */
    transmit();
    
    /* ------------------ Receive Messages ------------------ */
    receive(1);

    /* ------------------ Free Memory Space ------------------ */
    GC(clear)();

    /* ------------------ Close COM Port ------------------ */
    closePort();

    return 0;
}

crc32.c
/*-
 *  COPYRIGHT (C) 1986 Gary S. Brown.  You may use this program, or
 *  code or tables extracted from it, as desired without restriction.
 *
 *  First, the polynomial itself and its table of feedback terms.  The
 *  polynomial is
 *  X^32+X^26+X^23+X^22+X^16+X^12+X^11+X^10+X^8+X^7+X^5+X^4+X^2+X^1+X^0
 *
 *  Note that we take it "backwards" and put the highest-order term in
 *  the lowest-order bit.  The X^32 term is "implied"; the LSB is the
 *  X^31 term, etc.  The X^0 term (usually shown as "+1") results in
 *  the MSB being 1
 *
 *  Note that the usual hardware shift register implementation, which
 *  is what we're using (we're merely optimizing it by doing eight-bit
 *  chunks at a time) shifts bits into the lowest-order term.  In our
 *  implementation, that means shifting towards the right.  Why do we
 *  do it this way?  Because the calculated CRC must be transmitted in
 *  order from highest-order term to lowest-order term.  UARTs transmit
 *  characters in order from LSB to MSB.  By storing the CRC this way
 *  we hand it to the UART in the order low-byte to high-byte; the UART
 *  sends each low-bit to hight-bit; and the result is transmission bit
 *  by bit from highest- to lowest-order term without requiring any bit
 *  shuffling on our part.  Reception works similarly
 *
 *  The feedback terms table consists of 256, 32-bit entries.  Notes
 *
 *      The table can be generated at runtime if desired; code to do so
 *      is shown later.  It might not be obvious, but the feedback
 *      terms simply represent the results of eight shift/xor opera
 *      tions for all combinations of data and CRC register values
 *
 *      The values must be right-shifted by eight bits by the "updcrc
 *      logic; the shift must be unsigned (bring in zeroes).  On some
 *      hardware you could probably optimize the shift in assembler by
 *      using byte-swap instructions
 *      polynomial $edb88320
 *
 *
 * CRC32 code derived from work by Gary S. Brown.
 */

// #include <sys/param.h>
// #include <sys/systm.h>
#include <stdint.h>
// #include <stddef.h>

static uint32_t crc32_tab[] = {
	0x00000000, 0x77073096, 0xee0e612c, 0x990951ba, 0x076dc419, 0x706af48f,
	0xe963a535, 0x9e6495a3,	0x0edb8832, 0x79dcb8a4, 0xe0d5e91e, 0x97d2d988,
	0x09b64c2b, 0x7eb17cbd, 0xe7b82d07, 0x90bf1d91, 0x1db71064, 0x6ab020f2,
	0xf3b97148, 0x84be41de,	0x1adad47d, 0x6ddde4eb, 0xf4d4b551, 0x83d385c7,
	0x136c9856, 0x646ba8c0, 0xfd62f97a, 0x8a65c9ec,	0x14015c4f, 0x63066cd9,
	0xfa0f3d63, 0x8d080df5,	0x3b6e20c8, 0x4c69105e, 0xd56041e4, 0xa2677172,
	0x3c03e4d1, 0x4b04d447, 0xd20d85fd, 0xa50ab56b,	0x35b5a8fa, 0x42b2986c,
	0xdbbbc9d6, 0xacbcf940,	0x32d86ce3, 0x45df5c75, 0xdcd60dcf, 0xabd13d59,
	0x26d930ac, 0x51de003a, 0xc8d75180, 0xbfd06116, 0x21b4f4b5, 0x56b3c423,
	0xcfba9599, 0xb8bda50f, 0x2802b89e, 0x5f058808, 0xc60cd9b2, 0xb10be924,
	0x2f6f7c87, 0x58684c11, 0xc1611dab, 0xb6662d3d,	0x76dc4190, 0x01db7106,
	0x98d220bc, 0xefd5102a, 0x71b18589, 0x06b6b51f, 0x9fbfe4a5, 0xe8b8d433,
	0x7807c9a2, 0x0f00f934, 0x9609a88e, 0xe10e9818, 0x7f6a0dbb, 0x086d3d2d,
	0x91646c97, 0xe6635c01, 0x6b6b51f4, 0x1c6c6162, 0x856530d8, 0xf262004e,
	0x6c0695ed, 0x1b01a57b, 0x8208f4c1, 0xf50fc457, 0x65b0d9c6, 0x12b7e950,
	0x8bbeb8ea, 0xfcb9887c, 0x62dd1ddf, 0x15da2d49, 0x8cd37cf3, 0xfbd44c65,
	0x4db26158, 0x3ab551ce, 0xa3bc0074, 0xd4bb30e2, 0x4adfa541, 0x3dd895d7,
	0xa4d1c46d, 0xd3d6f4fb, 0x4369e96a, 0x346ed9fc, 0xad678846, 0xda60b8d0,
	0x44042d73, 0x33031de5, 0xaa0a4c5f, 0xdd0d7cc9, 0x5005713c, 0x270241aa,
	0xbe0b1010, 0xc90c2086, 0x5768b525, 0x206f85b3, 0xb966d409, 0xce61e49f,
	0x5edef90e, 0x29d9c998, 0xb0d09822, 0xc7d7a8b4, 0x59b33d17, 0x2eb40d81,
	0xb7bd5c3b, 0xc0ba6cad, 0xedb88320, 0x9abfb3b6, 0x03b6e20c, 0x74b1d29a,
	0xead54739, 0x9dd277af, 0x04db2615, 0x73dc1683, 0xe3630b12, 0x94643b84,
	0x0d6d6a3e, 0x7a6a5aa8, 0xe40ecf0b, 0x9309ff9d, 0x0a00ae27, 0x7d079eb1,
	0xf00f9344, 0x8708a3d2, 0x1e01f268, 0x6906c2fe, 0xf762575d, 0x806567cb,
	0x196c3671, 0x6e6b06e7, 0xfed41b76, 0x89d32be0, 0x10da7a5a, 0x67dd4acc,
	0xf9b9df6f, 0x8ebeeff9, 0x17b7be43, 0x60b08ed5, 0xd6d6a3e8, 0xa1d1937e,
	0x38d8c2c4, 0x4fdff252, 0xd1bb67f1, 0xa6bc5767, 0x3fb506dd, 0x48b2364b,
	0xd80d2bda, 0xaf0a1b4c, 0x36034af6, 0x41047a60, 0xdf60efc3, 0xa867df55,
	0x316e8eef, 0x4669be79, 0xcb61b38c, 0xbc66831a, 0x256fd2a0, 0x5268e236,
	0xcc0c7795, 0xbb0b4703, 0x220216b9, 0x5505262f, 0xc5ba3bbe, 0xb2bd0b28,
	0x2bb45a92, 0x5cb36a04, 0xc2d7ffa7, 0xb5d0cf31, 0x2cd99e8b, 0x5bdeae1d,
	0x9b64c2b0, 0xec63f226, 0x756aa39c, 0x026d930a, 0x9c0906a9, 0xeb0e363f,
	0x72076785, 0x05005713, 0x95bf4a82, 0xe2b87a14, 0x7bb12bae, 0x0cb61b38,
	0x92d28e9b, 0xe5d5be0d, 0x7cdcefb7, 0x0bdbdf21, 0x86d3d2d4, 0xf1d4e242,
	0x68ddb3f8, 0x1fda836e, 0x81be16cd, 0xf6b9265b, 0x6fb077e1, 0x18b74777,
	0x88085ae6, 0xff0f6a70, 0x66063bca, 0x11010b5c, 0x8f659eff, 0xf862ae69,
	0x616bffd3, 0x166ccf45, 0xa00ae278, 0xd70dd2ee, 0x4e048354, 0x3903b3c2,
	0xa7672661, 0xd06016f7, 0x4969474d, 0x3e6e77db, 0xaed16a4a, 0xd9d65adc,
	0x40df0b66, 0x37d83bf0, 0xa9bcae53, 0xdebb9ec5, 0x47b2cf7f, 0x30b5ffe9,
	0xbdbdf21c, 0xcabac28a, 0x53b39330, 0x24b4a3a6, 0xbad03605, 0xcdd70693,
	0x54de5729, 0x23d967bf, 0xb3667a2e, 0xc4614ab8, 0x5d681b02, 0x2a6f2b94,
	0xb40bbe37, 0xc30c8ea1, 0x5a05df1b, 0x2d02ef8d
};

uint32_t
crc32(uint32_t crc, const void *buf, size_t size)
{
	const uint8_t *p;

	p = buf;
	crc = crc ^ ~0U;

	while (size--)
		crc = crc32_tab[(crc ^ *p++) & 0xFF] ^ (crc >> 8);

	return crc ^ ~0U;
}

crc32.h
/*-
 *  COPYRIGHT (C) 1986 Gary S. Brown.  You may use this program, or
 *  code or tables extracted from it, as desired without restriction.
 *
 *  First, the polynomial itself and its table of feedback terms.  The
 *  polynomial is
 *  X^32+X^26+X^23+X^22+X^16+X^12+X^11+X^10+X^8+X^7+X^5+X^4+X^2+X^1+X^0
 *
 *  Note that we take it "backwards" and put the highest-order term in
 *  the lowest-order bit.  The X^32 term is "implied"; the LSB is the
 *  X^31 term, etc.  The X^0 term (usually shown as "+1") results in
 *  the MSB being 1
 *
 *  Note that the usual hardware shift register implementation, which
 *  is what we're using (we're merely optimizing it by doing eight-bit
 *  chunks at a time) shifts bits into the lowest-order term.  In our
 *  implementation, that means shifting towards the right.  Why do we
 *  do it this way?  Because the calculated CRC must be transmitted in
 *  order from highest-order term to lowest-order term.  UARTs transmit
 *  characters in order from LSB to MSB.  By storing the CRC this way
 *  we hand it to the UART in the order low-byte to high-byte; the UART
 *  sends each low-bit to hight-bit; and the result is transmission bit
 *  by bit from highest- to lowest-order term without requiring any bit
 *  shuffling on our part.  Reception works similarly
 *
 *  The feedback terms table consists of 256, 32-bit entries.  Notes
 *
 *      The table can be generated at runtime if desired; code to do so
 *      is shown later.  It might not be obvious, but the feedback
 *      terms simply represent the results of eight shift/xor opera
 *      tions for all combinations of data and CRC register values
 *
 *      The values must be right-shifted by eight bits by the "updcrc
 *      logic; the shift must be unsigned (bring in zeroes).  On some
 *      hardware you could probably optimize the shift in assembler by
 *      using byte-swap instructions
 *      polynomial $edb88320
 *
 *
 * CRC32 code derived from work by Gary S. Brown.
 */

#ifndef CRC32_H_
#define CRC32_H_
// #include <sys/param.h>
// #include <sys/systm.h>
#include <stdint.h>
#include <stddef.h>

static uint32_t crc32_tab[];

uint32_t crc32(uint32_t crc, const void *buf, size_t size);

#endif


serial.h
#ifndef _SERIAL_H
#define _SERIAL_H

#include <windows.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>


#define GC(type) gen_command_##type

/* ------------------ Serial environment parameter ------------------ */

#define RX_BUFFER_SIZE  256
// #define COM_PORT "\\\\.\\COM6"
// #define SERIAL_BAUDRATE 9600
#define SERIAL_BYTESIZE 8
#define SERIAL_STOPBITS 1
#define SERIAL_PARITY   0


/* ------------------ Structure of Command ------------------ */
struct command{
    uint8_t *body;    // main command
    size_t  length;   // length of command
    uint8_t lsb;      // 1 = LSB, 0 = MSB
    uint8_t ctr;      // counter of command_body, n_th byte of command
    uint8_t ctr_rst;  // if 1 will reset the counter
};
typedef struct command command_t;


/* ------------------ Global variables ------------------ */
extern command_t *_command_;
extern uint8_t status_;
extern HANDLE hComm_;


/* ------------------ Serial functions ------------------ */
void initComPort(uint8_t portNo, uint32_t baudrate);
void printCommand(void);
void transmit(void);
void receive(uint8_t print_format);
void closePort(void);

/* ------------------ Command generation functions ------------------ */
void gen_command_clear(void);
void gen_command_init(void);
void gen_command_hex(uint32_t cpn, size_t cpn_len);
void gen_command_ascii(uint8_t* cpn, size_t cpn_len);
void gen_command_crc(void);

#endif


serial.c
#include "serial.h"
#include "crc32.c"

command_t *_command_;
uint8_t status_;
HANDLE hComm_;


/* ================== Serial functions ================== */

void initComPort(uint8_t portNo, uint32_t baudrate){
/*
    @Discreption
        This function will open the serial COM port, and initialize the COM port. The bytesize, stopbit ,and parity bit is defined in serial.h. If there is error occured when calling this function, this program will be terminated.
    
    @Input
        portNo: number of COM port
        baudrate: baudrate of remote device
    
    @Ouput
        none
*/
    
/* ------------------ Initialize COM port ------------------ */
    uint8_t port[10];
    uint8_t tmp[2];

    itoa(portNo, tmp, 10);
    strcat(port, "\\\\.\\COM");
    strcat(port, tmp);

    //Open the serial com port
    hComm_ = CreateFile(port,                          //friendly name
                       GENERIC_READ | GENERIC_WRITE,   // Read/Write Access
                       0,              // No Sharing, ports cant be shared
                       NULL,           // No Security
                       OPEN_EXISTING,  // Open existing port only
                       0 ,             // Non Overlapped I/O
                       NULL);          // Null for Comm Devices

    if (hComm_ == INVALID_HANDLE_VALUE){
        fprintf(stderr,"opening serial port failed 0x%X\n",GetLastError());
        exit(0);
    }
    else printf("Serial port opened, ");


    /* --------------- Set Device Control Block structure --------------- */
    DCB SerialParams = { 0 };  // Initializing DCB structure
    SerialParams.fRtsControl = 0x00;
    SerialParams.DCBlength = sizeof(SerialParams);
    SerialParams.BaudRate = baudrate;
    SerialParams.ByteSize = SERIAL_BYTESIZE;
    SerialParams.StopBits = SERIAL_STOPBITS;
    SerialParams.Parity = SERIAL_PARITY;
    printf("baudrate = %d\n", SerialParams.BaudRate);
   

    /* --------------- Set receive mask --------------- */
    status_ = SetCommMask(hComm_, EV_RXCHAR);
    if (!status_) {
        fprintf(stderr,"SetCommMask failed 0x%X\n",GetLastError());
        exit(0);
    }
    else printf("SetCommMask success\n");

    /* --------------- Set comm status --------------- */
    status_ = SetCommState(hComm_, &SerialParams);
    if (!status_) {
        fprintf(stderr,"SetCommState failed 0x%X\n",GetLastError());
        exit(0);
    }
    else printf("SetCommState success\n");

    /* --------------- Get the current settings --------------- */        
    status_ = GetCommState(hComm_, &SerialParams);
    if (!status_) {
        fprintf(stderr,"GetCommState failed 0x%X\n",GetLastError());
        exit(0);
    }
    else printf("GetCommState success\n");

    
    /* --------------- Set timeout structure --------------- */
    COMMTIMEOUTS timeouts = { 0 };        // Initializing COMMTIMEOUTS structure
    timeouts.ReadIntervalTimeout = 50;       // timeout between bytes
    timeouts.ReadTotalTimeoutConstant = 500;  // total timeout in one transaction
    
    status_ = SetCommTimeouts(hComm_, &timeouts);
    if (!status_) {
        fprintf(stderr,"SetCommTimeouts failed 0x%X\n",GetLastError());
        exit(0);
    }
    else printf("SetCommTimeouts success\n");

}

void printCommand(void){
/*
    @Discreption
        This function will print the command to terminal.
    
    @Input
        none
    
    @Ouput
        none
*/

/* ------------------ Print for debug ------------------ */
    printf("\ncommand length = %d\n", _command_->length);
    printf("command = \n");

    for(int i = 0; i<_command_->length; i++){
        printf("%X ", *(_command_->body + i));
    }
    printf("\n");

    return;
}

void transmit(void){
/*
    @Discreption
        This function will transmit the command in `_command_->body`. If there is an error occured while transmittion, this program will be terminate to protect the remote device.
    
    @Input
        none
    
    @Ouput
        none
*/

/* ------------------ Send Command ------------------ */
    uint8_t BytesWritten = 0;
        
    status_ = WriteFile(hComm_,         // Handle to the Serial port
                    _command_->body,    // Data to be written to the port
                    _command_->length,  // No of bytes to write
                    &BytesWritten,      // Bytes written
                    NULL);
    
    /* --------------- Check transmit status --------------- */
    if (!status_) {
        fprintf(stderr,"\nWriteFile failed 0x%X\n",GetLastError());
        printf("Number of bytes written to the serial port = %d\n", BytesWritten);
        exit(0);
    }
    else printf("\nNumber of bytes written to the serial port = %d\n", BytesWritten);
    
    return;
}

void receive(uint8_t print_format){
/*
    @Discreption
        This function will receive the messages form the device, and move byte by byte to `rx_buffer` and the default buffer size depends on `RX_BUFFER_SIZE` defined in serial.h

        The terminate condition is receiving 0x0D (<CR> in ascii), or the `rx_buffer` is full.
    
    @Input
        none
    
    @Ouput
        none
*/

    /* --------------- Set rx buffer --------------- */
    uint8_t *rx_buffer = calloc(RX_BUFFER_SIZE, sizeof(uint8_t));
    
    /* ------------------ Receive Data ------------------ */
    uint32_t event_mask;    // Event mask to trigger
    // Setting WaitComm() Event
    status_ = WaitCommEvent(hComm_, &event_mask, NULL); //Wait for the character to be received

    /* --------------- Check receive status --------------- */
    if (!status_) {
        fprintf(stderr,"\nWaitCommEvent failed 0x%X\n",GetLastError());
        exit(0);
    }


    uint8_t rx_tmp;         // temperory Character
    uint16_t rx_ctr = 0;     // rx counter
    uint32_t BytesRead;     // Bytes read by ReadFile()

    /* --------------- Read data and store in a buffer --------------- */
    do{
        status_ = ReadFile(hComm_,          // Handle to the Serial port
                        &rx_tmp,            // read buffer 
                        sizeof(rx_tmp),     // Number of bytes to read
                        &BytesRead,         // Number of bytes read 
                        NULL);
        if (!status_) {
            fprintf(stderr,"\nReadFile failed 0x%X\n",GetLastError());
            exit(0);
        }
        else{
            rx_buffer[rx_ctr] = rx_tmp;
            rx_ctr++;
            // if ( rx_buffer[rx_ctr-1] == 0x0D ) break;
        }
    }
    while (BytesRead > 0 || rx_ctr >= RX_BUFFER_SIZE );
    // rx_ctr--; // Get Actual length of received data
    printf("\nNumber of bytes received = %d\n\n", rx_ctr);
    printf("Received messages = \n");
    
    /* ------------------ Print received data ------------------ */

    if (print_format == 1){
        for (int i = 0; i < rx_ctr; i++){
            printf("%c", rx_buffer[i]);
        }
    }
    else{
        for (int i = 0; i < rx_ctr; i++){
            printf("%X ", rx_buffer[i]);
        }
    }
    
    return;
}

void closePort(void){
/*
    @Discreption
        This function will close the COM port which just opened by calling iniComPort().
    
    @Input
        none
    
    @Ouput
        none
*/
    
/* ------------------ Close COM port ------------------ */
    status_ = CloseHandle(hComm_);     //Closing the Serial Port
    if (!status_) fprintf(stderr,"\nClosing Serial port failed 0x%X\n",GetLastError());
    else  printf("\nSerial port closed\n");
    return ;
}


/* ================== Command generation functions ================== */

void gen_command_hex(uint32_t cpn, size_t cpn_len){
/*
    @Discreption
        This function will split component into unit of bytes, and append it to command_body in command structure. And it will append component in MSB or LSB by "lsb" in command structure. If ctr_rst == 1, this function will set counter to 0
    
    @Input
        cmd: structure of command
        cpn: component which need to be appended to command, must less than 4 bytes
        cpn_len: length of component
    
    @Ouput
        cmd: structure of command
*/
    
    if (_command_->ctr_rst) _command_->ctr =0;
    if (_command_->lsb){
        for(int i=0; i<cpn_len; i++){
            *(_command_->body + _command_->ctr) = cpn >> 8*i;
            _command_->ctr++;
        }
    }
    else{        
        for(int i=cpn_len; i>0; i--){
            *(_command_->body + _command_->ctr) = cpn >> 8*(i-1);
            _command_->ctr++;
        }
    }    

    _command_->length += cpn_len;
    
    /* ----- for debug command use -----*/
    // for(int i = 0; i<_command_->length; i++){
    //     printf("%d = %X\n",i, *(_command_->body + i));
    // }

}

void gen_command_ascii(uint8_t* cpn, size_t cpn_len){
/*
    @Discreption
        This function will split component into unit of bytes, and append it to command_body in command structure. There is no MSB or LSB for ascii command. If ctr_rst == 1, this function will set counter to 0
    
    @Input
        cmd: structure of command
        cpn: component which need to be appended to command
        cpn_len: length of component
    
    @Ouput
        cmd: structure of command
*/
    
    if (_command_->ctr_rst) _command_->ctr =0;

    // /* to eliminate "\0" in string */
    // cpn_len -= 1;       

    for(int i=0; i<cpn_len; i++){
            *(_command_->body + _command_->ctr) = *(cpn+i);
            _command_->ctr++;
    }

    _command_->length += cpn_len;
    
    /* ----- for debug command use -----*/
    // for(int i = 0; i<_command_->length; i++){
    //     printf("%d = %X\n",i, *(_command_->body + i));
    // }

    return;
}

void gen_command_crc(void){
/*
    @Discreption
        This function will calculate the crc32 for the command, and use gen_command_hex to append the crc code to the command
    
    @Input
        cmd: structure of command
    
    @Ouput
        cmd: structure of command
*/

    uint32_t crc = 0x00000000; 
    crc = crc32(crc, _command_->body, _command_->length);     // get CRC32 from command
    gen_command_hex(crc, 4);
}

void gen_command_init(void){
/*
    @Discreption
        This function will create a memory space for command, and initialize the structure of command to zero
    
    @Input
        none
    
    @Ouput
        none
*/

    _command_ = (command_t*)malloc(sizeof(command_t));

    _command_->body = (uint8_t*)calloc(128, sizeof(uint8_t));
    _command_->length = 0;       // initial command length in byte
    _command_->lsb = 0;               // 0 = LSB, 1 = MSB
    _command_->ctr = 0;               // initial counter of command_body
    _command_->ctr_rst = 0;
    return;
}

void gen_command_clear(void){
/*
    @Discreption
        This function will free the memory space of command
    
    @Input
        none
    
    @Ouput
        none
*/

    free(_command_);
}
